src/equations/kdv_1d.jl

```diff
@@ -0,0 +1,183 @@
+    D::RealT # still-water depth
+    eta0::RealT # constant still-water surface
+end
+
+function KdVEquation1D(; gravity, D = 1.0, eta0 = 0.0)
+    eta0 == 0.0 || @warn "The still-water surface needs to be 0 for the KdV equations"
+    KdVEquation1D(gravity, D, eta0)
+end
+
+"""
+    initial_condition_convergence_test(x, t, equations::KdVEquation1D, mesh)
+
+A traveling-wave solution used for convergence tests in a periodic domain, here for dimensional variables.
+"""
+function initial_condition_convergence_test(x, t, equations::KdVEquation1D, mesh)
+    g = gravity(equations)
+    D = equations.D
+    c = 1.5 * sqrt(g * D)
+    A = 2 * D * (c - sqrt(g * D)) / sqrt(g * D)
+    K = 1 / 2 * sqrt(3 * A / D^3)
+    x_t = mod(x - c * t - xmin(mesh), xmax(mesh) - xmin(mesh)) + xmin(mesh)
+    eta = A * sech(K * x_t)^2
+    return SVector(eta)
+end
+
+"""
+    initial_condition_manufactured(x, t, equations::KdVEquation1D, mesh)
+
+A smooth manufactured solution in combination with [`source_terms_manufactured`](@ref).
+"""
+function initial_condition_manufactured(x, t, equations::KdVEquation1D,
+                                        mesh)
+    eta = 1 + exp(-t / 2) * sinpi(2 * (x - t / 2))
+    return SVector(eta)
+end
+
+"""
+    source_terms_manufactured(q, x, t, equations::KdVEquation1D, mesh)
+
+A smooth manufactured solution in combination with [`initial_condition_manufactured`](@ref).
+
+How it was calculated, is described in:
+https://github.com/NumericalMathematics/DispersiveShallowWater.jl/pull/198/files#r2090805751
+"""
+function source_terms_manufactured(q, x, t, equations::KdVEquation1D)
+    g = gravity(equations)
+    D = equations.D
+
+    a1 = sinpi(2x - t)
+    a2 = cospi(2x - t)
+    b1 = exp(-t / 2)
+    c0 = sqrt(g * D)
+    c1 = sqrt(g / D)
+
+    dq1 = -0.5 * a1 * b1 - pi * a2 * b1 +
+          2pi * a2 * c0 * b1 +
+          3pi * a2 * c1 * (1 + a1 * b1) * b1 -
+          (4 / 3) * D^2 * pi^3 * a2 * c0 * b1
+
+    return SVector(dq1)
+end
+
+function create_cache(mesh, equations::KdVEquation1D,
+                      solver, initial_condition,
+                      ::BoundaryConditionPeriodic,
+                      RealT, uEltype)
+    g = gravity(equations)
+    D = equations.D
+    DD = D^2
+    c_0 = sqrt(g * D)
+    c_1 = 0.5 * sqrt(g / D)
+
+    # We use `DiffCache` from PreallocationTools.jl to enable automatic/algorithmic differentiation
+    # via ForwardDiff.jl. 
+    # 1: eta
+    N = ForwardDiff.pickchunksize(1 * nnodes(mesh))
+    template = ones(RealT, nnodes(mesh))
+
+    eta2 = DiffCache(zero(template), N)
+    eta2_x = DiffCache(zero(template), N)
+
+    eta_x = DiffCache(zero(template), N)
+    eta_xxx = DiffCache(zero(template), N)
+
+    if solver.D1 isa PeriodicUpwindOperators
+        D1 = solver.D1.central
+
+        # calculate the third derivative operator using upwind operators
+        D3 = sparse(solver.D1.plus) * sparse(solver.D1.minus) * sparse(solver.D1.central)
```

Comment:
```suggestion
        D3 = sparse(solver.D1.plus) * sparse(solver.D1.central) * sparse(solver.D1.minus)
```
Don't we need the central one in the middle to prove semidiscrete energy conservation? I fear with this change, some test values may change a bit again. Sorry for missing that before :grimacing:

Reply:
ohh yes, this is a blunder on my side